src/ast/ast.h

```diff
@@ -30,103 +30,226 @@ typedef struct {
+);
+
+// returns the number of times the given clause appears in the AST
+uint AST_GetClauseCount
+(
+	const AST *ast,
+	cypher_astnode_type_t clause_type
+);
+
+// returns all instances of the given clause in the AST
+const cypher_astnode_t **AST_GetClauses
+(
+	const AST *ast,
+	cypher_astnode_type_t type
+);
+
+// returns an array (arr.h type) of all the nodes from a given type.
+// note: array must be free after use
+const cypher_astnode_t **AST_GetTypedNodes
+(
+	const cypher_astnode_t *ast,
+	cypher_astnode_type_t type
+);
+
+// collect all aliases within `entity` subtree
+void AST_CollectAliases
+(
+	const char ***aliases,
+	const cypher_astnode_t *entity
+);
+
+AST *AST_Build
+(
+	cypher_parse_result_t *parse_result
+);
+
+AST *AST_NewSegment
+(
+	AST *master_ast,
+	uint start_offset,
+	uint end_offset
+);
+
+// sets a parameter parsing result in the ast
+void AST_SetParamsParseResult
+(
+	AST *ast,
+	cypher_parse_result_t *params_parse_result
+);
+
+// returns a shallow copy of the original AST pointer with ref counter increased
+AST *AST_ShallowCopy
+(
+	AST *orig
+);
+
+// populate the AST's map of all referenced aliases
+void AST_BuildReferenceMap
+(
+	AST *ast,
+	const cypher_astnode_t *project_clause
+);
+
+// annotate AST naming all anonymous graph entities
+void AST_Enrich
+(
+	AST *ast
+);
+
+// returns true if the given alias is referenced within this AST segment
+bool AST_AliasIsReferenced
+(
+	AST *ast,
+	const char *alias
+);
+
+// returns true if the given identifier is used as an alias within this tree
+bool AST_IdentifierIsAlias
+(
+	const cypher_astnode_t *root,
+	const char *identifier
+);
+
+// convert an AST integer node (which is stored internally as a string) into an integer
+long AST_ParseIntegerNode
+(
+	const cypher_astnode_t *int_node
+);
+
+// returns true if the given clause contains an aggregate function
+bool AST_ClauseContainsAggregation
+(
+	const cypher_astnode_t *clause
+);
+
+// collect the aliases from a RETURN clause to populate ResultSet column names
+const char **AST_BuildReturnColumnNames
+(
+	const cypher_astnode_t *return_clause
+);
+
+// collect the aliases from a CALL clause to populate ResultSet column names
+const char **AST_BuildCallColumnNames
+(
+	const cypher_astnode_t *return_clause
+);
+
+// parse a query to construct an immutable AST
+cypher_parse_result_t *parse_query
+(
+	const char *query  // query to parse
+);
+
+// parse a query parameter values only
+// the remaining query string is set in the result body
+cypher_parse_result_t *parse_params
+(
+	const char *query,
+	const char **query_body
+);
+
+// free the immutable AST generated by the parser
+void parse_result_free
+(
+	cypher_parse_result_t *parse_result
+);
+
+// returns the ast annotation context collection of the AST
+AST_AnnotationCtxCollection *AST_GetAnnotationCtxCollection
+(
+	AST *ast
+);
+
+const char *AST_ToString
+(
+	const cypher_astnode_t *node
+);
+
+void AST_Free
+(
+	AST *ast
+);
```

Comment:
Just making sure, all changes in this file are simple formatting ?

Reply:
Yes